sessions: Migrate to custom indexed session model.

abdelrahman725 opened the pull request:
Fixes #19490.

[CZO discussion](https://chat.zulip.org/#narrow/stream/3-backend/topic/delete_user_sessions.20performance)

# Custom Session model indexed by (realm, user) and IP.

- New `RealmSession` model with fields `realm`, `user` and `ip_address`.
- Custom `SessionStore` to work with `RealmSession`.
- More performant  bulk deletion of sessions in `zerver/lib/sessions.py`
- Data migration file to copy over data from the old table `django_sessions` to the new table `zerver_realmsession`.
- This migration introduced one extra query in `zerver/tests/test_signup.py` and  `zerver/tests/test_home.py`.

alexmv reviewed (commented):
> Create RealmSession model with fields: realm, user and ip_address,
indexed by realm and user.

Indexing the IP address would also be useful.

> This change introduced one extra query for 'assert_database_query_count()'

Which query, and why?

It also looks like you'll need to add some tests for complete coverage.

abdelrahman725 commented on zerver/lib/sessions.py:
```diff
@@ -31,43 +29,37 @@ def get_session_dict_user(session_dict: Mapping[str, int]) -> Optional[int]:
-            continue
-        user_profile = get_user_profile_by_id(user_profile_id)
-        if not user_profile.is_active or user_profile.realm.deactivated:
+    for session in RealmSession.objects.filter(user__isnull=False).select_related("realm", "user"):
+        # Guaranteed to be UserProfile since we are excluding Null users from the queryset.
+        user_profile = cast(UserProfile, session.user)
+        # Guaranteed to be Realm since a user (at least in this session model) always has a realm.
+        user_realm = cast(Realm, session.realm)
+
+        if not user_profile.is_active or user_realm.deactivated:
```
Yes, that filter will also remove the need for the `if` condition checks.

abdelrahman725 commented on zerver/migrations/0507_sessions_new_custom_model.py:
```diff
@@ -0,0 +1,42 @@
+        ("zerver", "0506_realm_require_unique_names"),
+    ]
+
+    operations = [
+        migrations.CreateModel(
+            name="RealmSession",
+            fields=[
+                (
+                    "session_key",
+                    models.CharField(
+                        max_length=40, primary_key=True, serialize=False, verbose_name="session key"
+                    ),
+                ),
+                ("session_data", models.TextField(verbose_name="session data")),
+                ("expire_date", models.DateTimeField(db_index=True, verbose_name="expire date")),
+                ("ip_address", models.GenericIPAddressField(null=True)),
+                (
+                    "realm",
+                    models.ForeignKey(
+                        null=True, on_delete=django.db.models.deletion.CASCADE, to="zerver.realm"
+                    ),
+                ),
+                (
+                    "user",
+                    models.ForeignKey(
+                        null=True,
+                        on_delete=django.db.models.deletion.CASCADE,
+                        to=settings.AUTH_USER_MODEL,
+                    ),
+                ),
+            ],
+        ),
```
Where should this code (for copying over the data) live ? 

We want it be raw SQL query not django orm, right ?

abdelrahman725 replied:
1. After `INSERT INTO...`  query is successfull, should i delete the data from existing session table, `TRUNCATE TABLE django_session` ?

2. What the reason for prefering raw SQL over django orm, performance ?

abdelrahman725 commented on zerver/models/sessions.py:
```diff
@@ -0,0 +1,102 @@
+
+@receiver(pre_save, sender=RealmSession)
+def set_realm(*, instance: RealmSession, **kwargs: object) -> None:
+    if instance.user is not None:
+        instance.realm_id = instance.user.realm_id
```
When user logs out, that instance gets deleted

abdelrahman725 commented on zerver/views/home.py:
```diff
@@ -205,6 +205,8 @@ def home_real(request: HttpRequest) -> HttpResponse:
         # that we're a spectator, not a logged-in user.
         user_profile = None
 
+    request.session["realm_id"] = realm.id
```
Maybe needs more investigation, but i put it there based on this [suggestion](https://chat.zulip.org/#narrow/stream/3-backend/topic/delete_user_sessions.20performance/near/1770478) from CZO

alexmv replied:
You should have an argument for why _this_ change is the right change, not just that it seems to work.

abdelrahman725 commented on zerver/models/sessions.py:
```diff
@@ -0,0 +1,102 @@
+    @override
+    def create_model_instance(self, data: Dict[str, Union[int, str]]) -> RealmSession:
+        session_object = super().create_model_instance(data)
+        assert isinstance(session_object, RealmSession)
```
Yes, and in that case `assert isinstance` is the right way to silent that false alarm from `mypy` ?

alexmv replied:
In that case the right way to silence the false alarm is to _file a bug_ with the project whose types aren't correct.  We can then make this:
```python3
session_object: RealmSession = super().create_model_instance(data)  # type: ignore[assignment]  # link to bug report
```
...so we can know to come back and clean this up later.

abdelrahman725 commented:
> > This change introduced one extra query for 'assert_database_query_count()'
> 
> Which query, and why?

@alexmv 
```sql
UPDATE "zerver_realmsession" SET "session_data" = 'some value"',
 "expire_date" = 'whatever'::timestamptz, 
"realm_id" = 2, 
"user_id" = 10,
 "ip_address" = NULL WHERE "zerver_realmsession"."session_key" = 'some value'
```

That extra query is triggered because in `create_realm_session_store` called `zerver.models.create_model_instance()` we add new custom fields to `session_object` when we create or update it, These fields by default are part of `session_data`, so a new query is executed to update that `session_data`.

alexmv commented on zerver/migrations/0530_sessions_data_migration.py:
```diff
@@ -0,0 +1,108 @@
+# because creating an index on pre-existing data is quicker than updating it incrementally as each row is loaded.
+
+# Both drop indexes and copy data are done in one transaction
+# (because we can't write to postgres then update its scheme (i.e. create the indexes) in the same transaction),
```
Citation?  AFAIK you can insert data and alter the schema (not "scheme") in the same transaction.  In fact, dropping indexes is altering the schema, so by this argument you should need three transactions.

But I believe this is very much doable in one transaction.

abdelrahman725 replied:
> In fact, dropping indexes is altering the schema, so by this argument you should need three transactions.

Yes, but the following error

`OperationalError: cannot CREATE INDEX "zerver_realmsession" because it has pending trigger events`

is raised when creating indexes, not when dropping them.

alexmv commented on zerver/migrations/0530_sessions_data_migration.py:
```diff
@@ -0,0 +1,108 @@
+        with connection.cursor() as cursor:
+            # Fetch indexes names.
+            cursor.execute(
+                SQL("""SELECT indexname FROM pg_indexes WHERE tablename = 'zerver_realmsession' """)
```
I don't think the triple-quotes, or the `SQL(...)` even, are necessary here?

alexmv commented on zerver/migrations/0530_sessions_data_migration.py:
```diff
@@ -0,0 +1,108 @@
+                        """ALTER TABLE zerver_realmsession drop constraint zerver_realmsession_pkey"""
+                    )
+                else:
+                    if "session_key" in index[0]:
+                        session_key_index_name = index[0]
```
Why do we have this special case?

abdelrahman725 replied:
It was just a way for me to catch the generated index_name by django, which has session_key in it.

I removed this block anyway.

alexmv commented on zerver/migrations/0530_sessions_data_migration.py:
```diff
@@ -0,0 +1,108 @@
+                )
+            )
+
+            cursor.execute(
+                SQL(
+                    # {index_name} was auto generated by django, we create the index with the same name just in case,
+                    # as i'm not sure if a custom name in this case would be a problem.
+                    """
+                    CREATE INDEX IF NOT EXISTS {index_name}
+                        ON zerver_realmsession (session_key varchar_pattern_ops);
+
+                    CREATE INDEX IF NOT EXISTS zerver_realmsession_expire_date_idx
+                        ON zerver_realmsession (expire_date);
+
+                    CREATE INDEX IF NOT EXISTS zerver_realmsession_realm_idx
+                        ON zerver_realmsession (realm_id);
+
+                    CREATE INDEX IF NOT EXISTS zerver_realmsession_user_idx
+                        ON zerver_realmsession (user_id);
+
+                    CREATE INDEX IF NOT EXISTS zerver_realmsession_realm_user_idx
+                        ON zerver_realmsession (realm_id, user_id);
+
+                    CREATE INDEX IF NOT EXISTS zerver_realmsession_ip_address_idx
+                        ON zerver_realmsession (ip_address);
+                    """
+                ).format(index_name=Identifier(session_key_index_name))
```
We have the precise index definitions that we dropped, in the `indexdef` column:
```
zulip=> select * from pg_indexes where tablename = 'django_session';
 schemaname |   tablename    |                indexname                 | tablespace |                                                          indexdef
------------+----------------+------------------------------------------+------------+-----------------------------------------------------------------------------------------------------------------------------
 zulip      | django_session | django_session_session_key_c0390e0f_like |            | CREATE INDEX django_session_session_key_c0390e0f_like ON zulip.django_session USING btree (session_key varchar_pattern_ops)
 zulip      | django_session | django_session_pkey                      |            | CREATE UNIQUE INDEX django_session_pkey ON zulip.django_session USING btree (session_key)
 zulip      | django_session | django_session_expire_date_a5c62663      |            | CREATE INDEX django_session_expire_date_a5c62663 ON zulip.django_session USING btree (expire_date)
(3 rows)
```
I'd much rather not hard-code these, because it's very easy to accidentally have had drift.

Dropping the primary key constraint makes me leery -- I think my proposal is that we drop everything else, storing the `indexdef` statements as we go, and then re-run just those statements at the end.

abdelrahman725 replied:
I agree that we should avoid hard coding and go with executing those `indexdef`s instead.

But why are you leery of dropping the primary key ?

abdelrahman725 replied:
But why are you leery of dropping the primary key ?

alexmv commented on zerver/models/sessions.py:
```diff
@@ -0,0 +1,115 @@
+
+
+# This SessionsStore is not used in tests,
+# its none-cached version 'zerver/models/nocache_sessions.SessionStore' is the one used instead,
```
```suggestion
# its non-cached version 'zerver/models/nocache_sessions.SessionStore' is the one used instead,
```

alexmv commented on zerver/models/sessions.py:
```diff
@@ -0,0 +1,115 @@
+
+# This SessionsStore is not used in tests,
+# its none-cached version 'zerver/models/nocache_sessions.SessionStore' is the one used instead,
+# That class functions are tested in zerver/tests/test_sessions in test_sessionstore_create_model_instance
```
```suggestion
# Those class functions are tested in zerver/tests/test_sessions.py
```

abdelrahman725 commented on zerver/lib/sessions.py:
```diff
@@ -31,45 +31,54 @@ def get_session_dict_user(session_dict: Mapping[str, int]) -> Optional[int]:
+    # session caching is disabled during tests
+    if not settings.TEST_SUITE:
+        for session_key in session_keys:
+            session_engine.SessionStore(session_key).delete_cached_session()  # type: ignore[attr-defined] # not sure how to fix this
```
error: "SessionBase" has no attribute "delete_cached_session"  [attr-defined]

abdelrahman725 replied:
This is now becomes solvable after `post_delete()` signal approach as that block of code is now in `zerver.models.sessions` where `SessionStore` lives and is recognized.

abdelrahman725 commented on zerver/lib/sessions.py:
```diff
@@ -31,45 +31,54 @@ def get_session_dict_user(session_dict: Mapping[str, int]) -> Optional[int]:
-        if not user_profile.is_active or user_profile.realm.deactivated:
-            logging.info("Deactivating session for deactivated user %s", user_profile.id)
-            delete_session(session)
+    for session in RealmSession.objects.filter(
+        Q(user__is_active=False) | Q(realm__deactivated=True)
+    ):
+        # user_profile is guaranteed to NOT be None as None is excluded from the query.
+        user_profile = assert_is_not_none(session.user)
+        logging.info("Deactivating session for deactivated user %s", user_profile.id)
+        delete_session(session)
```
No, we would still need to delete the cached session too like others.

`delete_session()` is the only function we have that takes care of deleting the session from both th db and cache

abdelrahman725 replied:
Yes, `post_delete()` is more appropriate as the instance (session) key is still accessible to the signal.

I will go with `.delete()` and a signal approach, more clean and organized

timabbott commented on zerver/models/sessions.py:
```diff
@@ -0,0 +1,122 @@
+    class Meta:
+        indexes = [
+            models.Index(fields=("realm", "user"), name="zerver_realmsession_realm_user_idx"),
+            models.Index(fields=("ip_address",), name="zerver_realmsession_ip_address_idx"),
```
This one seems equivalent to setting `db_index=True` on the `ip_address` field?

abdelrahman725 replied:
Yes, when we want to just apply a single simple index with no  customization, `db_index=True` is more convenient

abdelrahman725 committed:
sessions: Migrate to custom indexed session model.

Fixes #19490.

Create RealmSession model with custom fields: realm, user and ip_address,
indexed by (realm, user) and ip_address.

Migrate code to use that model, the migrated code
interacts with sessions more efficiently, for example
bulk deletion of sessions in "zerver/lib/sessions".

Create data_migration to copy over data from
django_session table to zerver_realmsession table.
